// discloud-cli/build/util/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.app_version = exports.ModPermissions = exports.requiredDiscloudConfigProps = exports.mapDiscloudConfigProps = exports.required_files = exports.FileExt = exports.blocked_files = exports.locales = exports.configPath = exports.logsPath = exports.backupsPath = exports.cpu_arch = exports.os_platform = exports.os_release = exports.os_name = exports.version = void 0;
const gluegun_1 = require("@discloudapp/gluegun");
const node_fs_1 = require("node:fs");
const node_os_1 = require("node:os");
const node_path_1 = require("node:path");
exports.version = JSON.parse((0, node_fs_1.readFileSync)((0, node_path_1.join)(__dirname, "..", "..", "package.json"), "utf8")).version;
exports.os_name = (0, node_os_1.type)();
exports.os_release = (0, node_os_1.release)().split?.(".").slice(0, 2).join(".") ?? (0, node_os_1.release)();
exports.os_platform = (0, node_os_1.platform)();
exports.cpu_arch = (0, node_os_1.arch)();
exports.backupsPath = "discloud/backups";
exports.logsPath = "discloud/logs";
exports.configPath = `${gluegun_1.filesystem.homedir()}/.discloud`;
exports.locales = ["en-US", "pt-BR"];
exports.blocked_files = {
    common: [".git", ".vscode"],
    go: [],
    js: ["node_modules", "package-lock.json", "yarn.lock"],
    py: [],
    rb: ["Gemfile.lock"],
    rs: ["Cargo.lock", "target"],
    ts: ["node_modules", "package-lock.json", "yarn.lock"],
};
var FileExt;
(function (FileExt) {
    FileExt["cjs"] = "js";
    FileExt["cts"] = "ts";
    FileExt["go"] = "go";
    FileExt["js"] = "js";
    FileExt["jsx"] = "js";
    FileExt["mjs"] = "js";
    FileExt["mts"] = "ts";
    FileExt["py"] = "py";
    FileExt["rb"] = "rb";
    FileExt["rlib"] = "rs";
    FileExt["rs"] = "rs";
    FileExt["ts"] = "ts";
    FileExt["tsx"] = "ts";
})(FileExt = exports.FileExt || (exports.FileExt = {}));
exports.required_files = {
    common: ["discloud.config"],
    go: ["go.mod", "go.sum"],
    js: ["package.json"],
    py: ["requirements.txt"],
    rb: ["Gemfile"],
    rs: ["Cargo.toml"],
    ts: ["package.json"],
};
exports.mapDiscloudConfigProps = {
    site: {
        ID: "subdomain",
    },
};
exports.requiredDiscloudConfigProps = {
    bot: ["MAIN", "NAME", "TYPE", "RAM", "VERSION"],
    site: ["ID", "MAIN", "TYPE", "RAM", "VERSION"],
};
exports.ModPermissions = {
    backup_app: 1,
    commit_app: 2,
    edit_ram: 4,
    logs_app: 8,
    restart_app: 16,
    start_app: 32,
    status_app: 64,
    stop_app: 128,
};
exports.app_version = {
    go: ["latest"],
    java: ["latest"],
    js: ["latest", "current", "suja"],
    php: ["latest"],
    py: ["latest", "suja"],
    rb: ["latest"],
    rs: ["latest", "suja"],
};
