// discloud-cli/build/util/FsJson.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FsJson = void 0;
const gluegun_1 = require("@discloudapp/gluegun");
class FsJson {
    path;
    options;
    #data = {};
    constructor(path, options = {}) {
        this.path = path;
        this.options = options;
        this.options = {
            encoding: "utf8",
            ...this.options,
        };
        this.#data = this.#read() ?? this.#data;
    }
    get data() {
        return this.#data;
    }
    #encode(data = this.#data, encoding = this.options.encoding) {
        return Buffer.from(JSON.stringify(data)).toString(encoding);
    }
    #decode(path = this.path, encoding = this.options.encoding) {
        try {
            return JSON.parse(Buffer.from(gluegun_1.filesystem.read(path), encoding).toString("utf8"));
        }
        catch {
            return {};
        }
    }
    #read(path = this.path) {
        if (gluegun_1.filesystem.exists(path))
            try {
                return gluegun_1.filesystem.read(path, "json");
            }
            catch {
                return this.#decode(path);
            }
    }
    update(data, path = this.path) {
        this.#data = { ...this.#data, ...data };
        const encoded = this.#encode(this.#data);
        gluegun_1.filesystem.write(path, encoded);
        return this.#data;
    }
}
exports.FsJson = FsJson;
exports.default = FsJson;
