// discloud-cli/build/extensions/AskForApps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("@discloudapp/util");
const util_2 = require("../util");
function default_1(toolbox) {
    toolbox.prompt.askForApps = function (apps, options = {}) {
        options = {
            all: false,
            showStatus: true,
            ...options,
        };
        options.discloudConfig ??= new util_1.DiscloudConfig(options.discloudConfigPath);
        return toolbox.prompt.ask({
            name: "appId",
            message: "Choose the app",
            type: "select",
            choices: (options.all ? [{
                    name: "all",
                    message: "All apps",
                    value: "all",
                }] : []).concat((0, util_2.sortAppsBySameId)(apps, options.discloudConfig.data.ID).map(app => ({
                name: app.id,
                message: [
                    app.name, " - ", app.id,
                    options.showStatus ?
                        " - " + (app.online ?
                            toolbox.print.colors.green("online") :
                            toolbox.print.colors.red("offline")) :
                        "",
                    "perms" in app ?
                        app.perms.length ?
                            ` - [${app.perms.join()}]` :
                            "" :
                        "",
                ].join(""),
                value: app.id,
            }))),
            initial: options.all ? options.discloudConfig.data.ID ? apps.length ? 1 : 0 : 0 : 0,
        });
    };
}
exports.default = default_1;
