// discloud-cli/build/commands/apps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const v2_1 = require("@discloudapp/api-types/v2");
const util_1 = require("../util");
exports.default = {
    name: "apps",
    description: "Get information of your applications.",
    alias: ["app"],
    async run(toolbox) {
        const { print, parameters } = toolbox;
        if (!util_1.config.data.token)
            return print.error("Please use login command before using this command.");
        if (util_1.RateLimit.isLimited)
            return print.error(`Rate limited until: ${util_1.RateLimit.limited}`);
        const id = parameters.first || "all";
        const spin = print.spin({
            text: print.colors.cyan("Fetching apps..."),
        });
        const apiRes = await util_1.apidiscloud.get(v2_1.Routes.app(id));
        new util_1.RateLimit(apiRes.headers);
        print.spinApiRes(apiRes, spin, { exitOnError: true });
        if (!apiRes.data)
            return;
        if ("apps" in apiRes.data)
            print.table((0, util_1.makeTable)(apiRes.data.apps), {
                format: "lean",
            });
    },
};
