// discloud-cli/build/commands/locale.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const v2_1 = require("@discloudapp/api-types/v2");
const util_1 = require("../util");
const constants_1 = require("../util/constants");
exports.default = {
    name: "locale",
    description: "Set your locale.",
    async run(toolbox) {
        const { print, prompt } = toolbox;
        if (!util_1.config.data.token)
            return print.error("Please use login command before using this command.");
        if (util_1.RateLimit.isLimited)
            return print.error(`Rate limited until: ${util_1.RateLimit.limited}`);
        const { locale } = await prompt.ask({
            name: "locale",
            message: "Choose your locale",
            type: "select",
            choices: constants_1.locales,
        });
        const apiRes = await util_1.apidiscloud.put(v2_1.Routes.locale(locale), {});
        new util_1.RateLimit(apiRes.headers);
        print.printApiRes(apiRes);
        if (!apiRes.data)
            return;
        if ("localeList" in apiRes.data)
            print.info("Supported locales: " + apiRes.data.localeList?.join(", "));
    },
};
