// discloud-cli/build/commands/start.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const v2_1 = require("@discloudapp/api-types/v2");
const util_1 = require("../util");
exports.default = {
    name: "start",
    description: "Start one or all of your apps on Discloud.",
    alias: ["i"],
    async run(toolbox) {
        const { parameters, print, prompt } = toolbox;
        if (!util_1.config.data.token)
            return print.error("Please use login command before using this command.");
        if (util_1.RateLimit.isLimited)
            return print.error(`Rate limited until: ${util_1.RateLimit.limited}`);
        if (!parameters.first) {
            const { appId } = await prompt.fetchAndAskForApps({ all: true });
            if (!appId)
                return print.error("Need app id.");
            parameters.first = appId;
        }
        const spin = print.spin({
            text: print.colors.cyan("Starting..."),
        });
        const apiRes = await util_1.apidiscloud.put(v2_1.Routes.appStart(parameters.first), {});
        new util_1.RateLimit(apiRes.headers);
        print.spinApiRes(apiRes, spin, { exitOnError: true });
        if (!apiRes.data)
            return;
        if ("apps" in apiRes.data)
            print.table((0, util_1.makeTable)(apiRes.data.apps), {
                format: "lean",
            });
    },
};
